Restructure Cryptogram class + new class + tests

LaurenHimbeault opened the pull request:
Commit messages say it all, 

 fixed up the crypto class,
added a new class
added a ton of tests
did this with eric so four eyes have already gone over this but we will wait for someone other than us to do a final review here

Commit:
LOTS OF CHANGES HERE GUYS!

Added the Cryptogram Pair class to help me keep track of plaintext-ciphertext character pairs without using a hashmap
Fixed the cryptogram class so that it utilixes the new class, fixed the mapping structures according to our discussion at the last meeting
and I fixed the algorithms I use for encrypting and decrypting

I use a new algorithm called sattoros alg, i linked the url but look it up

also added some more tests and organized the current ones there

Commit:
More test cases

fixed isequals method

Commit:
More Tests

paymog commented on pg13/models/Cryptogram.java:
```diff
@@ -12,231 +12,289 @@
+	private final int ALPHABET_SIZE = 26;	// Letters of the alphabet
+	private String ciphertext;				// Coded string
+	private String plaintext;				// the original user string
+	private CryptogramPair[] charMapping;	// Mapping for plaintext to ciphertext characters used for encryption and decryption
```
Personally, I don't like the name "charMapping". Make something like "solutionMapping" might be better

paymog commented on pg13/models/Cryptogram.java:
```diff
@@ -12,231 +12,289 @@
-		this.ciphertext = encrypt(plaintext);
+		this.charMapping = setMappingKeys();
+		generateMappingKeys();
+		this.userMapping = setMappingKeys();
```
should we call this() instead of recalling these three lines?

paymog commented on pg13/models/Cryptogram.java:
```diff
@@ -12,231 +12,289 @@
 		return ciphertext;
 	}
 
+	public void setCiphertext(String ciphertext) {
```
Should this be private? I don't think it's great that consumers of the Cryptogram can choose custom ciphertexts

paymog commented on pg13/models/Cryptogram.java:
```diff
@@ -12,231 +12,289 @@
+	}
+
+	public void setPlaintext(String plaintext) {
+		this.plaintext = plaintext;
```
When the plaintext is updated, should the cipher text and hashes be updated?

paymog commented on pg13/models/Cryptogram.java:
```diff
@@ -12,231 +12,289 @@
+	}
+
+	public CryptogramPair[] getCharMapping() {
+		return charMapping;
```
Should this be returned defensively? Since arrays are passed by reference it might be safer to return a copy of the array as opposed to the array itself. Same with the getUserMapping()

paymog commented on pg13/models/Cryptogram.java:
```diff
@@ -12,231 +12,289 @@
+	 *  @paeam String userString
+	 *  @return checks if userString which is the user's plaintext, matches the original plaintext (ignorescase)
+	 */	
+	public boolean isCompleted(String userString)
```
Is this overriding the interface method? If not, that should probably be overridden.

paymog commented on pg13/models/CryptogramPair.java:
```diff
@@ -0,0 +1,43 @@
+	{
+		if(!isLetter(plainc) || !isLetter(cipherc))
+		{
+			throw new IllegalArgumentException();
```
You should include a message with the exception

paymog commented on pg13/models/Cryptogram.java:
```diff
@@ -12,231 +12,289 @@
-			}
-			//else it is punctuation or a number so we want to map it to itself
-			else
+	private char[] reOrderUserMappingByDecrypt() 
```
Make this and reOrderUserMappingByDecript into one method with a parameter of which CryptogramMapping[] to "fix"

Can we rename to something like SortMappingByCipherCharacter instead?

paymog commented on tests/models/CryptogramTest.java:
```diff
@@ -14,100 +15,134 @@ public CryptogramTest(String arg0)
 			assertNotNull(cryptogram);
 			assertEquals("Lauren Slusky", cryptogram.getAuthor());
 			assertEquals("Practice Cryptogram", cryptogram.getTitle());
-			System.out.println("Finished testCryptogramGeneralData");
```
assert the date and title too?

paymog commented on tests/models/CryptogramTest.java:
```diff
@@ -14,100 +15,134 @@ public CryptogramTest(String arg0)
-
-			System.out.println("\nStarting testCryptogramEmptyPlainText");		
+	
 			cryptogram = new Cryptogram("Lauren Slusky", "Practice Cryptogram", new Date(), "");
```
Since we use "Lauren Slusky", "Practice Cryptogram" and new Date() all the time, should we make these instance variables for the test?

paymog committed:
Test changes

I think the testing is a little more clean now.

I created some default variables so that we can pass the same thing
everytime instead of repeatedly typing ""Practice Cryptogram"" or "new
Date()"

I put in more assertions

willvoutier commented:
Is this good to go then?

willvoutier commented:
Just rebased the branch for an easier merge. Merging then.
